analyzers/src/SonarAnalyzer.Common/Rules/UnusedStringBuilderBase.cs

```diff
@@ -0,0 +1,64 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UnusedStringBuilderBase<TSyntaxKind, TVariableDeclarator, TInvocationExpression, TReturnStatement> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TVariableDeclarator : SyntaxNode
+    where TInvocationExpression : SyntaxNode
+    where TReturnStatement : SyntaxNode
+{
+    private const string DiagnosticId = "S3063";
+    protected override string MessageFormat => """Remove this "StringBuilder"; ".ToString()" is never called.""";
+
+    protected abstract string GetVariableName(TVariableDeclarator declaration);
+    protected abstract bool NeedsToTrack(TVariableDeclarator declaration, SemanticModel semanticModel);
+    protected abstract SyntaxNode GetAncestorBlock(TVariableDeclarator declaration);
+    protected abstract bool IsIsStringInvoked(string variableName, IList<TInvocationExpression> invocations);
+    protected abstract bool IsPassedToMethod(string variableName, IList<TInvocationExpression> invocations);
+    protected abstract bool IsReturned(string variableName, IList<TReturnStatement> returnStatements);
+
+    protected UnusedStringBuilderBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            var variableDeclaration = (TVariableDeclarator)c.Node;
+            if (!NeedsToTrack(variableDeclaration, c.SemanticModel))
+            {
+                return;
+            }
+            var variableName = GetVariableName(variableDeclaration);
```

Comment:
I think for CSharp the [ValueText property](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.syntaxtoken.valuetext?view=roslyn-dotnet-4.3.0) was doing the magic returning `sb` from `@sb` (it was already supported), not the case for VisualBasic case insensitive naming.
Anyway, I've changed the implementation using the semantic model all over the place and checking the semantic info.
Now we can support stuff like this:

```
public class MyClass
{
    void MyMethod()
    {
        StringBuilder sb = new(); // Noncompliant (before it was a FN)
        MyClass myClass = new();
        myClass.sb.ToString();
     }
}
public class MyClass
{
    public StringBuilder sb = new();
}
```

Reply:
Great. See https://github.com/SonarSource/sonar-dotnet/pull/6696#discussion_r1099086296 why we need to do the naming comparison anyway. Closing this because it is tracked on the other comment.